server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,164 @@
+    unsafeRunAsync: (() => F[ServerResponse[NettyResponse]]) => (Future[ServerResponse[NettyResponse]], () => Future[Unit]),
+    isSsl: Boolean
+)(implicit m: MonadError[F])
+    extends ChannelInboundHandlerAdapter {
+
+  // By using the Netty event loop assigned to this channel we get two benefits:
+  //  1. We can avoid the necessary hopping around of threads since Netty pipelines will
+  //     only pass events up and down from within the event loop to which it is assigned.
+  //     That means calls to ctx.read(), and ctx.write(..), would have to be trampolined otherwise.
+  //  2. We get serialization of execution: the EventLoop is a serial execution queue so
+  //     we can rest easy knowing that no two events will be executed in parallel.
+  private[this] var eventLoopContext: ExecutionContext = _
+
+  private val logger = LoggerFactory.getLogger(getClass.getName)
+
+  def isWsHandshake(req: HttpRequest): Boolean =
+    "Websocket".equalsIgnoreCase(req.headers().get(HttpHeaderNames.UPGRADE)) &&
+      "Upgrade".equalsIgnoreCase(req.headers().get(HttpHeaderNames.CONNECTION))
+
+  override def handlerAdded(ctx: ChannelHandlerContext): Unit =
+    if (ctx.channel.isActive) {
+      initHandler(ctx)
+    }
+  override def channelActive(ctx: ChannelHandlerContext): Unit = {
+    channelGroup.add(ctx.channel)
+    initHandler(ctx)
+  }
+
+  private[this] def initHandler(ctx: ChannelHandlerContext): Unit = {
+    if (eventLoopContext == null)
+      eventLoopContext = ExecutionContext.fromExecutor(ctx.channel.eventLoop)
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    logger.error("Error while processing the request", cause)
+  }
+
+  override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
+    def writeError500(req: HttpRequest, reason: Throwable): Unit = {
+      logger.error("Error while processing the request", reason)
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      val _ = ctx.writeAndFlush(res)
+    }
+    msg match {
+      case req: FullHttpRequest if isWsHandshake(req) =>
+        ctx.pipeline().remove(this)
+        ctx.pipeline().remove("readTimeoutHandler")
+        ReferenceCountUtil.release(msg)
+        val (runningFuture, _) = unsafeRunAsync { () =>
+          route(NettyServerRequest(req.retain()))
+            .map {
+              case Some(response) => response
+              case None           => ServerResponse.notFound
+            }
+        }
+
+        val _ = runningFuture.transform {
+          case Success(serverResponse) =>
+            try {
+              serverResponse.body match {
+                case Some(function) => {
+                  val content = function(ctx)
+                  content match {
+                    case r: ReactiveWebSocketProcessorNettyResponseContent => {
+                      ctx
```

Comment:
maybe move this to a dedicated function? this seems to be "the" place where we handle the WS logic, while all the other branches handle error conditions